File: python/weather.py
"""Weather Data Aggregator - Async API fetching with Python asyncio."""

import asyncio
import aiohttp
import json
import os
import time
from dataclasses import dataclass
from pathlib import Path
from typing import List, Dict, Optional, Protocol
from urllib.parse import quote

# Configuration Constants
REQUEST_TIMEOUT = aiohttp.ClientTimeout(total=10)


# Custom Exceptions
class WeatherError(Exception):
    """Base exception for weather-related errors."""

    pass


class GeocodingError(WeatherError):
    """Raised when city geocoding fails."""

    pass


class APIError(WeatherError):
    """Raised when API request fails."""

    pass


# Global weather code mappings (loaded once at startup like Go)
_WEATHER_CODES_PATH = Path(__file__).parent.parent / "weather_codes.json"
weather_code_mapping: dict = {}
condition_mapping: dict = {}


def load_weather_codes() -> None:
    """Load weather code mappings from JSON."""
    global weather_code_mapping, condition_mapping
    with open(_WEATHER_CODES_PATH) as f:
        codes = json.load(f)
    weather_code_mapping = codes
    condition_mapping = codes.get("conditions", {})


@dataclass
class WeatherData:
    """Weather from a single source with temperature, humidity, condition."""

    source: str
    temperature: float = 0.0
    humidity: Optional[float] = None
    condition: str = ""
    error: Optional[str] = None
    duration_ms: Optional[float] = None  # Request duration in milliseconds


class WeatherSource(Protocol):
    """Weather source interface - implements fetch(city, session, coords_cache)."""

    name: str

    async def fetch(
        self,
        city: str,
        session: aiohttp.ClientSession,
        coords_cache: Optional[Dict[str, tuple[float, float]]] = None,
    ) -> WeatherData: ...


def init_sources() -> list:
    """Initialize weather sources (free + API-key sources if configured)."""
    sources = [OpenMeteoSource()]

    # Add API-key sources if keys are available
    if key := os.getenv("TOMORROW_API_KEY"):
        sources.append(TomorrowIOSource(key))

    if key := os.getenv("WEATHER_API_COM_KEY"):
        sources.append(WeatherAPISource(key))

    if key := os.getenv("METEOSOURCE_API_KEY"):
        sources.append(MeteosourceSource(key))

    if key := os.getenv("PIRATE_WEATHER_API_KEY"):
        sources.append(PirateWeatherSource(key))

    return sources


def normalize_source(name: str) -> str:
    """Normalize source name: lowercase, remove spaces/dashes/dots."""
    return name.replace(" ", "").replace("-", "").replace(".", "").lower()


async def http_get_json(url: str, session: aiohttp.ClientSession) -> dict:
    """HTTP GET returning JSON data. Raises APIError on failure.

    Args:
        url: URL to fetch
        session: aiohttp client session

    Returns:
        Parsed JSON response as dictionary

    Raises:
        APIError: If request fails (timeout, network, HTTP error, invalid JSON)
    """
    try:
        async with session.get(url) as resp:
            if resp.status != 200:
                raise APIError(f"HTTP {resp.status}")
            return await resp.json()
    except asyncio.TimeoutError:
        raise APIError("timeout")
    except json.JSONDecodeError as e:
        raise APIError(f"invalid JSON: {str(e)}")
    except aiohttp.ClientError as e:
        raise APIError(f"request failed: {str(e)}")


async def geocode_city(
    city: str, session: aiohttp.ClientSession
) -> tuple[float, float]:
    """Resolve city name to (lat, lon) using Open-Meteo geocoding API.

    Args:
        city: City name to geocode
        session: aiohttp client session

    Returns:
        Tuple of (latitude, longitude)

    Raises:
        GeocodingError: If geocoding fails or city not found
    """
    url = f"https://geocoding-api.open-meteo.com/v1/search?name={quote(city)}&count=1"
    try:
        data = await http_get_json(url, session)
    except APIError as e:
        raise GeocodingError(f"geocoding request failed: {str(e)}") from e

    if not data or not data.get("results"):
        raise GeocodingError("city not found")

    result = data["results"][0]
    return (result["latitude"], result["longitude"])


async def get_coordinates(
    city: str,
    session: aiohttp.ClientSession,
    coords_cache: Optional[Dict[str, tuple[float, float]]] = None,
) -> tuple[float, float]:
    """Get coords from cache or geocode if not cached.

    Returns cached coordinates or performs geocoding.
    Raises GeocodingError if geocoding fails.
    """
    if coords_cache and city in coords_cache:
        return coords_cache[city]
    return await geocode_city(city, session)


class BaseAPISource:
    """Base class for API sources requiring authentication."""

    def __init__(self, api_key: str):
        self._api_key = api_key


class OpenMeteoSource:
    """Open-Meteo API - no key required."""

    name = "Open-Meteo"

    async def fetch(
        self,
        city: str,
        session: aiohttp.ClientSession,
        coords_cache: Optional[Dict[str, tuple[float, float]]] = None,
    ) -> WeatherData:
        result = WeatherData(source=self.name)

        try:
            coords = await get_coordinates(city, session, coords_cache)
            lat, lon = coords

            url = (
                f"https://api.open-meteo.com/v1/forecast?"
                f"latitude={lat}&longitude={lon}&"
                f"current=temperature_2m,relative_humidity_2m,weather_code"
            )

            data = await http_get_json(url, session)
            current = data.get("current", {})

            result.temperature = float(current.get("temperature_2m", 0))
            result.humidity = (
                float(current.get("relative_humidity_2m"))
                if current.get("relative_humidity_2m") is not None
                else None
            )
            result.condition = _map_wmo_code(current.get("weather_code", 0))

        except (GeocodingError, APIError) as e:
            result.error = str(e)
        except (ValueError, TypeError, KeyError) as e:
            result.error = f"data parsing error: {str(e)}"

        return result


class TomorrowIOSource(BaseAPISource):
    """Tomorrow.io API - requires key, coordinate-based."""

    name = "Tomorrow.io"

    async def fetch(
        self,
        city: str,
        session: aiohttp.ClientSession,
        coords_cache: Optional[Dict[str, tuple[float, float]]] = None,
    ) -> WeatherData:
        result = WeatherData(source=self.name)
        if not self._api_key:
            result.error = "API key required"
            return result

        try:
            coords = await get_coordinates(city, session, coords_cache)
            lat, lon = coords

            url = f"https://api.tomorrow.io/v4/weather/realtime?location={lat},{lon}&apikey={self._api_key}"
            data = await http_get_json(url, session)

            data_obj = data.get("data", {})
            values = data_obj.get("values", {})

            result.temperature = float(values.get("temperature", 0))
            result.humidity = (
                float(values.get("humidity"))
                if values.get("humidity") is not None
                else None
            )
            result.condition = values.get("weatherCode", "")
            if result.condition:
                result.condition = _map_tomorrow_code(result.condition)

        except (GeocodingError, APIError) as e:
            result.error = str(e)
        except (ValueError, TypeError, KeyError) as e:
            result.error = f"data parsing error: {str(e)}"

        return result


class WeatherAPISource(BaseAPISource):
    """WeatherAPI.com - requires key."""

    name = "WeatherAPI.com"

    async def fetch(
        self,
        city: str,
        session: aiohttp.ClientSession,
        coords_cache: Optional[Dict[str, tuple[float, float]]] = None,
    ) -> WeatherData:
        result = WeatherData(source=self.name)
        if not self._api_key:
            result.error = "API key required"
            return result

        try:
            url = f"https://api.weatherapi.com/v1/current.json?key={self._api_key}&q={quote(city)}"
            data = await http_get_json(url, session)

            current = data.get("current", {})
            result.temperature = float(current.get("temp_c", 0))
            result.humidity = (
                float(current.get("humidity"))
                if current.get("humidity") is not None
                else None
            )
            result.condition = current.get("condition", {}).get("text", "")

        except APIError as e:
            result.error = str(e)
        except (ValueError, TypeError, KeyError) as e:
            result.error = f"data parsing error: {str(e)}"

        return result


class MeteosourceSource(BaseAPISource):
    """Meteosource API - requires key, no available humidity on free tier."""

    name = "Meteosource"

    async def fetch(
        self,
        city: str,
        session: aiohttp.ClientSession,
        coords_cache: Optional[Dict[str, tuple[float, float]]] = None,
    ) -> WeatherData:
        result = WeatherData(source=self.name)
        if not self._api_key:
            result.error = "API key required"
            return result

        try:
            coords = await get_coordinates(city, session, coords_cache)
            lat, lon = coords

            url = (
                f"https://www.meteosource.com/api/v1/free/point?"
                f"lat={lat}&lon={lon}&sections=current&"
                f"language=en&units=metric&key={self._api_key}"
            )
            data = await http_get_json(url, session)

            current = data.get("current", {})
            result.temperature = float(current.get("temperature", 0))
            result.condition = current.get("summary", "")
            result.humidity = (
                float(current.get("humidity"))
                if current.get("humidity") is not None
                else None
            )

        except (GeocodingError, APIError) as e:
            result.error = str(e)
        except (ValueError, TypeError, KeyError) as e:
            result.error = f"data parsing error: {str(e)}"

        return result


class PirateWeatherSource(BaseAPISource):
    """Pirate Weather API - requires key."""

    name = "Pirate-Weather"

    async def fetch(
        self,
        city: str,
        session: aiohttp.ClientSession,
        coords_cache: Optional[Dict[str, tuple[float, float]]] = None,
    ) -> WeatherData:
        result = WeatherData(source=self.name)
        if not self._api_key:
            result.error = "API key required"
            return result

        try:
            coords = await get_coordinates(city, session, coords_cache)
            lat, lon = coords

            url = f"https://api.pirateweather.net/forecast/{self._api_key}/{lat},{lon}?units=si"
            data = await http_get_json(url, session)

            currently = data.get("currently", {})
            result.temperature = float(currently.get("temperature", 0))

            humidity_raw = currently.get("humidity")
            result.humidity = (
                float(humidity_raw) * 100 if humidity_raw and humidity_raw > 0 else None
            )
            result.condition = currently.get("summary", "")

        except (GeocodingError, APIError) as e:
            result.error = str(e)
        except (ValueError, TypeError, KeyError) as e:
            result.error = f"data parsing error: {str(e)}"

        return result


async def _fetch_with_timing(
    source: WeatherSource,
    city: str,
    session: aiohttp.ClientSession,
    coords_cache: Dict[str, tuple[float, float]],
) -> WeatherData:
    """Fetch weather from a source and measure duration."""
    start = time.perf_counter()
    result = await source.fetch(city, session, coords_cache)
    result.duration_ms = (time.perf_counter() - start) * 1000
    return result


async def fetch_weather_concurrently(
    city: str, sources: List[WeatherSource]
) -> List[WeatherData]:
    """Fetch from all sources concurrently using asyncio.gather."""
    async with aiohttp.ClientSession(timeout=REQUEST_TIMEOUT) as session:
        coords_cache: Dict[str, tuple[float, float]] = {}

        try:
            coords = await geocode_city(city, session)
            coords_cache[city] = coords
        except GeocodingError:
            pass
        tasks = [
            _fetch_with_timing(source, city, session, coords_cache)
            for source in sources
        ]
        return await asyncio.gather(*tasks)


async def fetch_weather_sequentially(
    city: str, sources: List[WeatherSource]
) -> List[WeatherData]:
    """Fetch weather sequentially for performance comparison."""
    async with aiohttp.ClientSession(timeout=REQUEST_TIMEOUT) as session:
        coords_cache: Dict[str, tuple[float, float]] = {}

        try:
            coords = await geocode_city(city, session)
            coords_cache[city] = coords
        except GeocodingError:
            pass

        return [
            await _fetch_with_timing(source, city, session, coords_cache)
            for source in sources
        ]


def aggregate_weather(data: List[WeatherData]) -> Dict:
    """Calculate avg temp/humidity and consensus condition from valid data."""
    result = {
        "avg_temp": 0.0,
        "avg_hum": 0.0,
        "hum_count": 0,
        "condition": "No data",
        "valid_count": 0,
    }

    if not data:
        return result

    valid = [d for d in data if d.error is None]
    if not valid:
        result["condition"] = "No valid data"
        return result

    result["valid_count"] = len(valid)
    result["avg_temp"] = sum(d.temperature for d in valid) / len(valid)

    humidities = [d.humidity for d in valid if d.humidity is not None]
    if humidities:
        result["avg_hum"] = sum(humidities) / len(humidities)
        result["hum_count"] = len(humidities)

    condition_counts: Dict[str, int] = {}
    for d in valid:
        normalized = normalize_condition(d.condition)
        condition_counts[normalized] = condition_counts.get(normalized, 0) + 1

    if condition_counts:
        result["condition"] = max(condition_counts, key=condition_counts.get)
    else:
        result["condition"] = "Unknown"

    return result


def _map_wmo_code(code: int) -> str:
    """Map WMO weather codes to readable conditions."""
    for range_def in weather_code_mapping["wmo"]["ranges"]:
        if range_def["min"] <= code <= range_def["max"]:
            return range_def["condition"]
    return "Unknown"


def _map_tomorrow_code(code: str) -> str:
    """Map Tomorrow.io codes to readable conditions."""
    try:
        code_str = str(int(code))
    except (ValueError, TypeError):
        return "Unknown"
    return weather_code_mapping["tomorrow_io"].get(code_str, "Unknown")


def normalize_condition(condition: str) -> str:
    """Normalize conditions to standard categories via keyword matching.

    Checks more specific patterns first (e.g., 'Partly Cloudy' before 'Cloudy').
    """
    lower = condition.lower()

    # Check in priority order (most specific first)
    condition_order = [
        "Partly Cloudy",
        "Clear",
        "Cloudy",
        "Rainy",
        "Snowy",
        "Foggy",
        "Stormy",
    ]

    for normalized in condition_order:
        if normalized in condition_mapping:
            if any(kw in lower for kw in condition_mapping[normalized]["keywords"]):
                return normalized

    return condition


def get_condition_emoji(condition: str) -> str:
    """Map condition to emoji. Returns thermometer if no match."""
    lower = condition.lower()
    for normalized, info in condition_mapping.items():
        if normalized.lower() in lower:
            return info["emoji"]
        if any(kw in lower for kw in info["keywords"]):
            return info["emoji"]
    return "🌡️"

File: go/weather.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// WeatherCodeRange represents a range of weather codes mapped to a condition.
type WeatherCodeRange struct {
	Min       int    `json:"min"`
	Max       int    `json:"max"`
	Condition string `json:"condition"`
}

// WeatherCodeConfig represents the structure of weather_codes.json.
type WeatherCodeConfig struct {
	WMO struct {
		Ranges []WeatherCodeRange `json:"ranges"`
	} `json:"wmo"`
	TomorrowIO map[string]string `json:"tomorrow_io"`
	Conditions map[string]struct {
		Keywords []string `json:"keywords"`
		Emoji    string   `json:"emoji"`
	} `json:"conditions"`
}

// WeatherCodes holds the unified weather code mappings loaded from JSON.
var WeatherCodes WeatherCodeConfig
var weatherCodesOnce sync.Once
var weatherCodesErr error


// WeatherData represents weather from a single source.
// Temperature in Celsius, Humidity as percentage (0-100).
type WeatherData struct {
	Source      string
	Temperature float64
	Humidity    *float64 // Pointer to distinguish between 0% and missing data
	Condition   string
	Error       error
	Duration    time.Duration
}

// WeatherSource interface - each source implements Fetch().
type WeatherSource interface {
	Fetch(ctx context.Context, city string, coordsCache map[string][2]float64) WeatherData
	Name() string
}


// client is a shared HTTP client with 10s timeout.
var client = &http.Client{
	Timeout: 10 * time.Second,
}

// loadWeatherCodes loads weather code mappings from shared JSON file.
// Uses sync.Once to ensure thread-safe single initialization.
func loadWeatherCodes() error {
	weatherCodesOnce.Do(func() {
		path := filepath.Join("..", "weather_codes.json")
		data, err := os.ReadFile(path)
		if err != nil {
			weatherCodesErr = fmt.Errorf("failed to read weather_codes.json: %w", err)
			return
		}
		if err := json.Unmarshal(data, &WeatherCodes); err != nil {
			weatherCodesErr = fmt.Errorf("failed to parse weather_codes.json: %w", err)
			return
		}
	})
	return weatherCodesErr
}


// doGet creates request with context and returns response + duration.
func doGet(ctx context.Context, url string) (*http.Response, time.Duration, error) {
	start := time.Now()
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, time.Since(start), fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("User-Agent", "weather-aggregator/1.0")

	resp, err := client.Do(req)
	duration := time.Since(start)
	if err != nil {
		return nil, duration, fmt.Errorf("request failed: %w", err)
	}
	if resp.StatusCode != http.StatusOK {
		resp.Body.Close()
		return nil, duration, fmt.Errorf("HTTP %d: %s", resp.StatusCode, resp.Status)
	}
	return resp, duration, nil
}

// getCoordinates gets coordinates from cache or performs geocoding.
func getCoordinates(ctx context.Context, city string, coordsCache map[string][2]float64) (float64, float64, error) {
	if coordsCache != nil {
		if coords, ok := coordsCache[city]; ok {
			return coords[0], coords[1], nil
		}
	}
	return lookupLatLon(ctx, city)
}

// lookupLatLon resolves a city name to coordinates using Open-Meteo geocoding.
func lookupLatLon(ctx context.Context, city string) (float64, float64, error) {
	geoURL := fmt.Sprintf("https://geocoding-api.open-meteo.com/v1/search?name=%s&count=1", url.QueryEscape(city))
	resp, _, err := doGet(ctx, geoURL)
	if err != nil {
		return 0, 0, fmt.Errorf("geocoding request failed: %w", err)
	}
	defer resp.Body.Close()

	var geo struct {
		Results []struct {
			Lat float64 `json:"latitude"`
			Lon float64 `json:"longitude"`
		} `json:"results"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&geo); err != nil {
		return 0, 0, fmt.Errorf("failed to decode geocoding response: %w", err)
	}
	if len(geo.Results) == 0 {
		return 0, 0, fmt.Errorf("city %q not found", city)
	}
	return geo.Results[0].Lat, geo.Results[0].Lon, nil
}

// --- Weather API Implementations ---
// Each API source implements the WeatherSource interface.
// Free sources: Open-Meteo
// API key required: WeatherAPI.com, Weatherstack, Meteosource, Pirate Weather, Tomorrow.io

// OpenMeteoSource - free API, no key required.
type OpenMeteoSource struct{}

func (o *OpenMeteoSource) Name() string { return "Open-Meteo" }
func (o *OpenMeteoSource) Fetch(ctx context.Context, city string, coordsCache map[string][2]float64) WeatherData {
	start := time.Now()
	res := WeatherData{Source: o.Name()}

	// Ensure weather codes are loaded
	if err := loadWeatherCodes(); err != nil {
		res.Error = fmt.Errorf("configuration error: %w", err)
		res.Duration = time.Since(start)
		return res
	}

	lat, lon, err := getCoordinates(ctx, city, coordsCache)
	if err != nil {
		res.Error = err
		res.Duration = time.Since(start)
		return res
	}

	weatherURL := fmt.Sprintf("https://api.open-meteo.com/v1/forecast?latitude=%.4f&longitude=%.4f&current=temperature_2m,relative_humidity_2m,weather_code", lat, lon)
	resp, _, err := doGet(ctx, weatherURL)
	if err != nil {
		res.Error = fmt.Errorf("weather request failed: %w", err)
		res.Duration = time.Since(start)
		return res
	}
	defer resp.Body.Close()

	var data struct {
		Current struct {
			Temp float64 `json:"temperature_2m"`
			Hum  float64 `json:"relative_humidity_2m"`
			Code int     `json:"weather_code"`
		}
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		res.Error = fmt.Errorf("failed to decode weather response: %w", err)
		res.Duration = time.Since(start)
		return res
	}
	res.Temperature = data.Current.Temp
	hum := data.Current.Hum
	res.Humidity = &hum
	res.Condition = mapWMOCode(data.Current.Code)
	res.Duration = time.Since(start)
	return res
}

// TomorrowIOSource - requires API key, coordinate-based.
type TomorrowIOSource struct {
	apiKey string
}

func (t *TomorrowIOSource) Name() string { return "Tomorrow.io" }
func (t *TomorrowIOSource) Fetch(ctx context.Context, city string, coordsCache map[string][2]float64) WeatherData {
	start := time.Now()
	res := WeatherData{Source: t.Name()}

	if t.apiKey == "" {
		res.Error = fmt.Errorf("API key required")
		res.Duration = time.Since(start)
		return res
	}

	lat, lon, err := getCoordinates(ctx, city, coordsCache)
	if err != nil {
		res.Error = err
		res.Duration = time.Since(start)
		return res
	}

	url := fmt.Sprintf("https://api.tomorrow.io/v4/weather/realtime?location=%.4f,%.4f&apikey=%s", lat, lon, t.apiKey)
	resp, _, err := doGet(ctx, url)
	if err != nil {
		res.Error = fmt.Errorf("weather request failed: %w", err)
		res.Duration = time.Since(start)
		return res
	}
	defer resp.Body.Close()

	var data struct {
		Data struct {
			Values struct {
				Temp      float64 `json:"temperature"`
				Hum       float64 `json:"humidity"`
				WeatherCd int     `json:"weatherCode"`
			} `json:"values"`
		} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		res.Error = fmt.Errorf("failed to decode response: %w", err)
		res.Duration = time.Since(start)
		return res
	}

	res.Temperature = data.Data.Values.Temp
	hum := data.Data.Values.Hum
	res.Humidity = &hum
	res.Condition = mapTomorrowCode(data.Data.Values.WeatherCd)
	res.Duration = time.Since(start)
	return res
}

// WeatherAPISource - requires API key.
type WeatherAPISource struct{ key string }

func (w *WeatherAPISource) Name() string { return "WeatherAPI.com" }
func (w *WeatherAPISource) Fetch(ctx context.Context, city string, coordsCache map[string][2]float64) WeatherData {
	start := time.Now()
	res := WeatherData{Source: w.Name()}
	if w.key == "" {
		res.Error = fmt.Errorf("API key required")
		res.Duration = time.Since(start)
		return res
	}
	resp, _, err := doGet(ctx, fmt.Sprintf("https://api.weatherapi.com/v1/current.json?key=%s&q=%s", w.key, url.QueryEscape(city)))
	if err != nil {
		res.Error = fmt.Errorf("weather request failed: %w", err)
		res.Duration = time.Since(start)
		return res
	}
	defer resp.Body.Close()
	var data struct {
		Current struct {
			TempC float64 `json:"temp_c"`
			Hum   float64 `json:"humidity"`
			Cond  struct {
				Text string `json:"text"`
			} `json:"condition"`
		} `json:"current"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		res.Error = fmt.Errorf("failed to decode response: %w", err)
		res.Duration = time.Since(start)
		return res
	}
	res.Temperature = data.Current.TempC
	hum := data.Current.Hum
	res.Humidity = &hum
	res.Condition = data.Current.Cond.Text
	res.Duration = time.Since(start)
	return res
}

// WeatherstackSource - requires API key, HTTP only on free tier.
type WeatherstackSource struct{ key string }

func (w *WeatherstackSource) Name() string { return "Weatherstack" }
func (w *WeatherstackSource) Fetch(ctx context.Context, city string, coordsCache map[string][2]float64) WeatherData {
	start := time.Now()
	res := WeatherData{Source: w.Name()}
	if w.key == "" {
		res.Error = fmt.Errorf("API key required")
		res.Duration = time.Since(start)
		return res
	}
	resp, _, err := doGet(ctx, fmt.Sprintf("http://api.weatherstack.com/current?access_key=%s&query=%s", w.key, url.QueryEscape(city)))
	if err != nil {
		res.Error = fmt.Errorf("weather request failed: %w", err)
		res.Duration = time.Since(start)
		return res
	}
	defer resp.Body.Close()
	var data struct {
		Current struct {
			Temp float64  `json:"temperature"`
			Hum  int      `json:"humidity"`
			Desc []string `json:"weather_descriptions"`
		} `json:"current"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		res.Error = fmt.Errorf("failed to decode response: %w", err)
		res.Duration = time.Since(start)
		return res
	}
	res.Temperature = data.Current.Temp
	hum := float64(data.Current.Hum)
	res.Humidity = &hum
	if len(data.Current.Desc) > 0 {
		res.Condition = data.Current.Desc[0]
	}
	res.Duration = time.Since(start)
	return res
}

// MeteosourceSource - requires API key, may lack humidity on free tier.
type MeteosourceSource struct{ key string }

func (m *MeteosourceSource) Name() string { return "Meteosource" }
func (m *MeteosourceSource) Fetch(ctx context.Context, city string, coordsCache map[string][2]float64) WeatherData {
	start := time.Now()
	res := WeatherData{Source: m.Name()}
	if m.key == "" {
		res.Error = fmt.Errorf("API key required")
		res.Duration = time.Since(start)
		return res
	}
	lat, lon, err := getCoordinates(ctx, city, coordsCache)
	if err != nil {
		res.Error = err
		res.Duration = time.Since(start)
		return res
	}
	resp, _, err := doGet(ctx, fmt.Sprintf("https://www.meteosource.com/api/v1/free/point?lat=%.4f&lon=%.4f&sections=current&language=en&units=metric&key=%s", lat, lon, m.key))
	if err != nil {
		res.Error = fmt.Errorf("weather request failed: %w", err)
		res.Duration = time.Since(start)
		return res
	}
	defer resp.Body.Close()
	var data struct {
		Current struct {
			Temp    float64     `json:"temperature"`
			Hum     interface{} `json:"humidity"`
			Summary string      `json:"summary"`
		} `json:"current"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		res.Error = fmt.Errorf("failed to decode response: %w", err)
		res.Duration = time.Since(start)
		return res
	}
	res.Temperature, res.Condition = data.Current.Temp, data.Current.Summary
	if h, ok := data.Current.Hum.(float64); ok {
		res.Humidity = &h
	} else if s, ok := data.Current.Hum.(string); ok {
		var hum float64
		fmt.Sscanf(strings.TrimSuffix(s, "%"), "%f", &hum)
		res.Humidity = &hum
	}
	res.Duration = time.Since(start)
	return res
}

// PirateWeatherSource - Dark Sky compatible, requires API key.
type PirateWeatherSource struct{ key string }

func (p *PirateWeatherSource) Name() string { return "Pirate Weather" }

func (p *PirateWeatherSource) Fetch(ctx context.Context, city string, coordsCache map[string][2]float64) WeatherData {
	start := time.Now()
	res := WeatherData{Source: p.Name()}
	if p.key == "" {
		res.Error = fmt.Errorf("API key required")
		res.Duration = time.Since(start)
		return res
	}
	lat, lon, err := getCoordinates(ctx, city, coordsCache)
	if err != nil {
		res.Error = err
		res.Duration = time.Since(start)
		return res
	}
	resp, _, err := doGet(ctx, fmt.Sprintf("https://api.pirateweather.net/forecast/%s/%.4f,%.4f?units=si", p.key, lat, lon))
	if err != nil {
		res.Error = fmt.Errorf("weather request failed: %w", err)
		res.Duration = time.Since(start)
		return res
	}
	defer resp.Body.Close()
	var data struct {
		Currently struct {
			Temp float64 `json:"temperature"`
			Hum  float64 `json:"humidity"`
			Sum  string  `json:"summary"`
		} `json:"currently"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		res.Error = fmt.Errorf("failed to decode response: %w", err)
		res.Duration = time.Since(start)
		return res
	}
	res.Temperature = data.Currently.Temp
	hum := data.Currently.Hum * 100
	res.Humidity = &hum
	res.Condition = data.Currently.Sum
	res.Duration = time.Since(start)
	return res
}

// fetchWeatherConcurrently fetches from all sources in parallel using goroutines.
// Pre-geocodes the city to reduce redundant API calls.
func fetchWeatherConcurrently(ctx context.Context, city string, sources []WeatherSource) []WeatherData {
	// Pre-geocode city once to avoid redundant calls from each source
	coordsCache := make(map[string][2]float64)
	if lat, lon, err := lookupLatLon(ctx, city); err == nil {
		coordsCache[city] = [2]float64{lat, lon}
	}

	ch := make(chan WeatherData, len(sources))
	for _, s := range sources {
		go func(src WeatherSource) { ch <- src.Fetch(ctx, city, coordsCache) }(s)
	}
	results := make([]WeatherData, 0, len(sources))
	for i := 0; i < len(sources); i++ {
		results = append(results, <-ch)
	}
	return results
}

// fetchSequential fetches weather data sequentially for performance comparison.
func fetchSequential(ctx context.Context, city string, sources []WeatherSource) []WeatherData {
	// Pre-geocode city once to avoid redundant calls
	coordsCache := make(map[string][2]float64)
	if lat, lon, err := lookupLatLon(ctx, city); err == nil {
		coordsCache[city] = [2]float64{lat, lon}
	}

	results := make([]WeatherData, 0, len(sources))
	for _, s := range sources {
		results = append(results, s.Fetch(ctx, city, coordsCache))
	}
	return results
}

// AggregateWeather calculates avg temp/humidity and consensus condition from valid data.
func AggregateWeather(data []WeatherData) (avgTemp, avgHum float64, cond string, valid int) {
	if len(data) == 0 {
		return 0, 0, "No data", 0
	}

	var tempSum, humSum float64
	var humCount int
	condCount := make(map[string]int)

	for _, d := range data {
		if d.Error == nil {
			tempSum += d.Temperature
			if d.Humidity != nil {
				humSum += *d.Humidity
				humCount++
			}
			condCount[normalizeCondition(d.Condition)]++
			valid++
		}
	}

	if valid == 0 {
		return 0, 0, "No valid data", 0
	}

	avgTemp = tempSum / float64(valid)
	if humCount > 0 {
		avgHum = humSum / float64(humCount)
	}

	maxCount := 0
	for c, count := range condCount {
		if count > maxCount {
			maxCount, cond = count, c
		}
	}
	return
}


// mapWMOCode converts WMO codes to readable conditions.
func mapWMOCode(code int) string {
	for _, r := range WeatherCodes.WMO.Ranges {
		if code >= r.Min && code <= r.Max {
			return r.Condition
		}
	}
	return "Unknown"
}

// mapTomorrowCode converts Tomorrow.io codes to readable conditions.
func mapTomorrowCode(code int) string {
	if condition := WeatherCodes.TomorrowIO[fmt.Sprintf("%d", code)]; condition != "" {
		return condition
	}
	return "Unknown"
}

// normalizeCondition converts conditions to standard categories.
// Checks more specific patterns first (e.g., "Partly Cloudy" before "Cloudy").
func normalizeCondition(c string) string {
	lower := strings.ToLower(c)
	
	// Check in priority order (most specific first)
	conditionOrder := []string{"Partly Cloudy", "Clear", "Cloudy", "Rainy", "Snowy", "Foggy", "Stormy"}
	
	for _, normalized := range conditionOrder {
		if info, exists := WeatherCodes.Conditions[normalized]; exists {
			for _, keyword := range info.Keywords {
				if strings.Contains(lower, keyword) {
					return normalized
				}
			}
		}
	}
	return c
}

// GetConditionEmoji maps conditions to emoji. Returns thermometer if no match.
func GetConditionEmoji(c string) string {
	lower := strings.ToLower(c)
	for _, info := range WeatherCodes.Conditions {
		for _, keyword := range info.Keywords {
			if strings.Contains(lower, keyword) {
				return info.Emoji
			}
		}
	}
	return "🌡️"
}
